Panic when inserting decimal to DB

<!--
If you are asking a question and are not reporting a bug or requesting a feature
please use the new Github Discussions feature we've enabled.
-->

If you're having a generation problem please answer these questions before submitting your issue. Thanks!

### What version of SQLBoiler are you using (`sqlboiler --version`)?
SQLBoiler v4.6.0

### What is your database and version (eg. Postgresql 10)
MySQL 5.7


### If this happened at runtime what code produced the issue? (if not applicable leave blank)

1. Create a table with a decimal NOT NULL column and generate the model.
2. Insert a new row with the column set to the zero value `types.Decimal{}`

```go
	// CREATE TABLE test
	//(
	//    id      int AUTO_INCREMENT,
	//    balance decimal NOT NULL,
	//    CONSTRAINT test_pk
	//        PRIMARY KEY (id)
	//);

	row := models.Test{Balance: types.Decimal{}}
	db, err := sql.Open("mysql", "root:@(localhost:3306)/test")
	if err != nil {
		panic(err)
	}
	boil.SetDB(db)

	row.Insert(boil.GetDB(), boil.Infer())
```

3. A panic occurs:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0x11b12bc]

goroutine 1 [running]:
github.com/ericlagergren/decimal.(*Big).IsNaN(...)
        /Users/shaoyanghong/GolandProjects/awesomeProject/vendor/github.com/ericlagergren/decimal/big.go:853
github.com/volatiletech/sqlboiler/v4/types.decimalValue(0x0, 0x14b6700, 0x0, 0x0, 0x8b95068, 0x14b7cc0)
        /Users/shaoyanghong/GolandProjects/awesomeProject/vendor/github.com/volatiletech/sqlboiler/v4/types/decimal.go:175 +0x3c
github.com/volatiletech/sqlboiler/v4/types.Decimal.Value(0x0, 0xc000135550, 0x100cd65, 0x129ace0, 0x12d58e0)
        /Users/shaoyanghong/GolandProjects/awesomeProject/vendor/github.com/volatiletech/sqlboiler/v4/types/decimal.go:61 +0x30
github.com/go-sql-driver/mysql.callValuerValue(0x8b95068, 0x0, 0x0, 0x8b95068, 0x0, 0x1)
        /Users/shaoyanghong/GolandProjects/awesomeProject/vendor/github.com/go-sql-driver/mysql/statement.go:203 +0xbe
github.com/go-sql-driver/mysql.converter.ConvertValue(0x12d58e0, 0x0, 0xc0002040c0, 0x12e9018, 0xc000080000, 0x8b95028)
        /Users/shaoyanghong/GolandProjects/awesomeProject/vendor/github.com/go-sql-driver/mysql/statement.go:145 +0x726
github.com/go-sql-driver/mysql.(*mysqlConn).CheckNamedValue(0xc00011b7a0, 0xc0002040c0, 0xc000135601, 0x8b95028)
        /Users/shaoyanghong/GolandProjects/awesomeProject/vendor/github.com/go-sql-driver/mysql/connection.go:639 +0x38
database/sql.driverArgsConnLocked(0x133e100, 0xc00011b7a0, 0x0, 0xc000013060, 0x1, 0x1, 0x2d, 0x12d0920, 0x100bc8c, 0x1485f60, ...)
        /usr/local/go/src/database/sql/convert.go:177 +0x190
database/sql.(*DB).execDC.func2()
        /usr/local/go/src/database/sql/sql.go:1573 +0xee
database/sql.withLock(0x133cfa0, 0xc000210000, 0xc000135a40)
        /usr/local/go/src/database/sql/sql.go:3294 +0x69
database/sql.(*DB).execDC(0xc00007fd40, 0x133e8b0, 0xc00001a0a0, 0xc000210000, 0xc000135b68, 0xc00001c420, 0x29, 0xc000013060, 0x1, 0x1, ...)
        /usr/local/go/src/database/sql/sql.go:1572 +0x495
database/sql.(*DB).exec(0xc00007fd40, 0x133e8b0, 0xc00001a0a0, 0xc00001c420, 0x29, 0xc000013060, 0x1, 0x1, 0x1, 0x1520438, ...)
        /usr/local/go/src/database/sql/sql.go:1557 +0x13c
database/sql.(*DB).ExecContext(0xc00007fd40, 0x133e8b0, 0xc00001a0a0, 0xc00001c420, 0x29, 0xc000013060, 0x1, 0x1, 0xc00005aa08, 0x199, ...)
        /usr/local/go/src/database/sql/sql.go:1535 +0xe5
database/sql.(*DB).Exec(0xc00007fd40, 0xc00001c420, 0x29, 0xc000013060, 0x1, 0x1, 0xc000013060, 0x1, 0x1, 0x6)
        /usr/local/go/src/database/sql/sql.go:1549 +0x85
awesomeProject/models.(*Test).Insert(0xc00005aa00, 0x133e0a0, 0xc00007fd40, 0x1, 0x0, 0x0, 0x0, 0x0, 0xffffffff)
        /Users/shaoyanghong/GolandProjects/awesomeProject/models/test.go:303 +0x33b
main.main()
        /Users/shaoyanghong/GolandProjects/awesomeProject/main.go:28 +0x13d
```



### What is the output of the command above with the `-d` flag added to it? (Provided you are comfortable sharing this, it contains a blueprint of your schema)


### Please provide a relevant database schema so we can replicate your issue (Provided you are comfortable sharing this)

	CREATE TABLE test
	(
	   id      int AUTO_INCREMENT,
	   balance decimal NOT NULL,
	   CONSTRAINT test_pk
	       PRIMARY KEY (id)
	);


### Further information. What did you do, what did you expect?

--- a/types/decimal.go
+++ b/types/decimal.go
@@ -167,8 +167,8 @@ func randomDecimal(nextInt func() int64, fieldType string, shouldBeNull bool) *d
 	return random
 }
 
-func decimalValue(d *decimal.Big, canNull bool) (driver.Value, error) {
-	if canNull && d == nil {
+func decimalValue(d *decimal.Big, canNil bool) (driver.Value, error) {
+	if !canNil && d == nil {
 		return nil, nil
 	}
 

--- a/types/decimal_test.go
+++ b/types/decimal_test.go
@@ -37,6 +37,10 @@ func TestDecimal_Value(t *testing.T) {
 		}
 	}
 
+	zero := Decimal{}
+	if _, err := zero.Value(); err != nil {
+		t.Error("zero value should not error")
+	}
 	infinity := Decimal{Big: new(decimal.Big).SetInf(true)}
 	if _, err := infinity.Value(); err == nil {
 		t.Error("infinity should not be passed into the database")

